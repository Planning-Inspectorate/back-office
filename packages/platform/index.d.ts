import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
