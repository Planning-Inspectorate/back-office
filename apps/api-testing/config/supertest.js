// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
