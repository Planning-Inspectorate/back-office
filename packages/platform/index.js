export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
