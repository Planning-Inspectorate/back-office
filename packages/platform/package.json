{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
