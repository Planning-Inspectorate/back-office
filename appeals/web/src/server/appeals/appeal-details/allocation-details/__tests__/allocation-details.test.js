import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { allocationDetailsData } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details/1/allocation-details';

describe('Allocation Details', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	it(`should render "Select the allocation level" Page`, async () => {
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);

		const response = await request.get(`${baseUrl}/allocation-level`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
	});

	it('should render "Select the allocation level" with error (no answer provided)', async () => {
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);

		const response = await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': ''
		});

		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
	});

	it(`should redirect to "Allocation specialism" page if no errors are present and posted allocation-level is provided`, async () => {
		const response = await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': 'A'
		});

		expect(response.statusCode).toBe(302);
	});

	it(`should render "Select allocation specialism" Page`, async () => {
		// post to preceding pages in the flow is necessary to set required data in the session
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);
		nock('http://test/')
			.get('/appeals/appeal-allocation-specialisms')
			.reply(200, allocationDetailsData.specialisms);

		await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': 'A'
		});

		const response = await request.get(`${baseUrl}/allocation-specialism`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
	});

	it('should render "Select allocation specialism" with error (no answer provided)', async () => {
		// post to preceding pages in the flow is necessary to set required data in the session
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);
		nock('http://test/')
			.get('/appeals/appeal-allocation-specialisms')
			.reply(200, allocationDetailsData.specialisms);

		await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': 'A'
		});

		const response = await request.post(`${baseUrl}/allocation-specialism`).send({
			'allocation-specialisms': []
		});

		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
	});

	it(`should redirect to "Check answer" page if no errors are present and posted allocation-specialisms is provided`, async () => {
		// post to preceding pages in the flow is necessary to set required data in the session
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);
		nock('http://test/')
			.get('/appeals/appeal-allocation-specialisms')
			.reply(200, allocationDetailsData.specialisms);

		await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': 'A'
		});

		const response = await request.post(`${baseUrl}/allocation-specialism`).send({
			'allocation-specialisms': [1, 2, 3]
		});

		expect(response.statusCode).toBe(302);
	});

	it(`should render "Check answers" Page`, async () => {
		// post to preceding pages in the flow is necessary to set required data in the session
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);
		nock('http://test/')
			.get('/appeals/appeal-allocation-specialisms')
			.reply(200, allocationDetailsData.specialisms);

		await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': 'A'
		});

		await request.post(`${baseUrl}/allocation-specialism`).send({
			'allocation-specialisms': [1, 2, 3]
		});

		const response = await request.get(`${baseUrl}/check-answers`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
	});

	it(`should redirect to "Case details" page if no errors are present and posted allocation-specialisms is provided`, async () => {
		// post to preceding pages in the flow is necessary to set required data in the session
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels);
		nock('http://test/')
			.get('/appeals/appeal-allocation-specialisms')
			.reply(200, allocationDetailsData.specialisms);
		nock('http://test/').patch('/appeals/1/appeal-allocation').reply(200);

		await request.post(`${baseUrl}/allocation-level`).send({
			'allocation-level': 'A'
		});

		await request.post(`${baseUrl}/allocation-specialism`).send({
			'allocation-specialisms': [1, 2, 3]
		});

		const response = await request.post(`${baseUrl}/check-answers`);

		expect(response.statusCode).toBe(302);
	});
});
