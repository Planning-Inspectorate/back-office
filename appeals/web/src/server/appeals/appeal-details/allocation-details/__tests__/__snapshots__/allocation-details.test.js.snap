// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allocation Details should render "Check answers" Page 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
            <h1             class="govuk-heading-xl govuk-!-margin-bottom-3">Check answers</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds govuk-!-margin-top-5 govuk-!-margin-bottom-6">
            <table class="govuk-table">
                <tbody class="govuk-table__body">
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Level</strong>
                        </td>
                        <td class="govuk-table__cell">A</td>
                        <td class="govuk-table__cell"><a href="/appeals-service/appeal-details/1/allocation-details/allocation-level">Change</a>
                        </td>
                    </tr>
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Band</strong>
                        </td>
                        <td class="govuk-table__cell">3</td>
                        <td class="govuk-table__cell"></td>
                    </tr>
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Specialism</strong>
                        </td>
                        <td class="govuk-table__cell">
                            <ul class="govuk-!-padding-0 govuk-!-margin-0 govuk-!-margin-left-4">
                                <li>Specialism 1</li>
                                <li>Specialism 2</li>
                                <li>Specialism 3</li>
                            </ul>
                        </td>
                        <td class="govuk-table__cell"><a href="/appeals-service/appeal-details/1/allocation-details/allocation-specialism">Change</a>
                        </td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>
    <form action="" method="POST" novalidate>
        <button type="submit" class="govuk-button" data-module="govuk-button">Confirm</button>
    </form>
</main>"
`;

exports[`Allocation Details should render "Select allocation specialism" Page 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
            <h1             class="govuk-heading-xl govuk-!-margin-bottom-3">Allocation specialism</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds govuk-!-margin-top-5 govuk-!-margin-bottom-6">
            <table class="govuk-table">
                <tbody class="govuk-table__body">
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Level</strong>
                        </td>
                        <td class="govuk-table__cell">A</td>
                    </tr>
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Band</strong>
                        </td>
                        <td class="govuk-table__cell">3</td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full">
            <form action="" method="POST" novalidate>
                <div class="govuk-form-group">
                    <fieldset class="govuk-fieldset">
                        <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
                            <h1 class="govuk-fieldset__heading"> Select allocation specialism</h1>
                        </legend>
                        <div class="govuk-checkboxes govuk-checkboxes--small" data-module="govuk-checkboxes">
                            <div class="govuk-checkboxes__item">
                                <input class="govuk-checkboxes__input" id="allocation-specialisms" name="allocation-specialisms"
                                type="checkbox" value="1">
                                <label class="govuk-label govuk-checkboxes__label" for="allocation-specialisms">Specialism 1</label>
                            </div>
                            <div class="govuk-checkboxes__item">
                                <input class="govuk-checkboxes__input" id="allocation-specialisms-2" name="allocation-specialisms"
                                type="checkbox" value="2">
                                <label class="govuk-label govuk-checkboxes__label" for="allocation-specialisms-2">Specialism 2</label>
                            </div>
                            <div class="govuk-checkboxes__item">
                                <input class="govuk-checkboxes__input" id="allocation-specialisms-3" name="allocation-specialisms"
                                type="checkbox" value="3">
                                <label class="govuk-label govuk-checkboxes__label" for="allocation-specialisms-3">Specialism 3</label>
                            </div>
                        </div>
                    </fieldset>
                </div>
                <button class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`Allocation Details should render "Select allocation specialism" with error (no answer provided) 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#allocation-specialisms">Please select one or more allocation specialisms</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
            <h1             class="govuk-heading-xl govuk-!-margin-bottom-3">Allocation specialism</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds govuk-!-margin-top-5 govuk-!-margin-bottom-6">
            <table class="govuk-table">
                <tbody class="govuk-table__body">
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Level</strong>
                        </td>
                        <td class="govuk-table__cell">A</td>
                    </tr>
                    <tr class="govuk-table__row">
                        <td class="govuk-table__cell"><strong>Band</strong>
                        </td>
                        <td class="govuk-table__cell">3</td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full">
            <form action="" method="POST" novalidate>
                <div class="govuk-form-group">
                    <fieldset class="govuk-fieldset">
                        <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
                            <h1 class="govuk-fieldset__heading"> Select allocation specialism</h1>
                        </legend>
                        <div class="govuk-checkboxes govuk-checkboxes--small" data-module="govuk-checkboxes">
                            <div class="govuk-checkboxes__item">
                                <input class="govuk-checkboxes__input" id="allocation-specialisms" name="allocation-specialisms"
                                type="checkbox" value="1">
                                <label class="govuk-label govuk-checkboxes__label" for="allocation-specialisms">Specialism 1</label>
                            </div>
                            <div class="govuk-checkboxes__item">
                                <input class="govuk-checkboxes__input" id="allocation-specialisms-2" name="allocation-specialisms"
                                type="checkbox" value="2">
                                <label class="govuk-label govuk-checkboxes__label" for="allocation-specialisms-2">Specialism 2</label>
                            </div>
                            <div class="govuk-checkboxes__item">
                                <input class="govuk-checkboxes__input" id="allocation-specialisms-3" name="allocation-specialisms"
                                type="checkbox" value="3">
                                <label class="govuk-label govuk-checkboxes__label" for="allocation-specialisms-3">Specialism 3</label>
                            </div>
                        </div>
                    </fieldset>
                </div>
                <button class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`Allocation Details should render "Select the allocation level" Page 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
            <h1             class="govuk-heading-xl govuk-!-margin-bottom-3">Select the allocation level</h1>
        </div>
    </div>
    <div class="govuk-!-margin-top-8">
        <div class="govuk-grid-row">
            <div class="govuk-grid-column-full">
                <form action="" method="POST" novalidate>
                    <div class="govuk-form-group">
                        <div class="govuk-radios" data-module="govuk-radios">
                            <div class="govuk-radios__item">
                                <input class="govuk-radios__input" id="allocation-level" name="allocation-level"
                                type="radio" value="A">
                                <label class="govuk-label govuk-radios__label" for="allocation-level">A</label>
                            </div>
                            <div class="govuk-radios__item">
                                <input class="govuk-radios__input" id="allocation-level-2" name="allocation-level"
                                type="radio" value="B">
                                <label class="govuk-label govuk-radios__label" for="allocation-level-2">B</label>
                            </div>
                        </div>
                    </div>
                    <button class="govuk-button" data-module="govuk-button">Continue</button>
                </form>
            </div>
        </div>
    </div>
</main>"
`;

exports[`Allocation Details should render "Select the allocation level" with error (no answer provided) 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#allocation-level">Allocation level must be provided</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
            <h1             class="govuk-heading-xl govuk-!-margin-bottom-3">Select the allocation level</h1>
        </div>
    </div>
    <div class="govuk-!-margin-top-8">
        <div class="govuk-grid-row">
            <div class="govuk-grid-column-full">
                <form action="" method="POST" novalidate>
                    <div class="govuk-form-group">
                        <div class="govuk-radios" data-module="govuk-radios">
                            <div class="govuk-radios__item">
                                <input class="govuk-radios__input" id="allocation-level" name="allocation-level"
                                type="radio" value="A">
                                <label class="govuk-label govuk-radios__label" for="allocation-level">A</label>
                            </div>
                            <div class="govuk-radios__item">
                                <input class="govuk-radios__input" id="allocation-level-2" name="allocation-level"
                                type="radio" value="B">
                                <label class="govuk-label govuk-radios__label" for="allocation-level-2">B</label>
                            </div>
                        </div>
                    </div>
                    <button class="govuk-button" data-module="govuk-button">Continue</button>
                </form>
            </div>
        </div>
    </div>
</main>"
`;
